k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +109,277 @@ class SingleNsNamer(
   }
 }
 
+object SingleNsNamer {
+  protected val PrefixLen = 2
+}
+
 abstract class EndpointsNamer(
   idPrefix: Path,
-  mkApi: String => NsApi,
-  backoff: Stream[Duration] = Backoff.exponentialJittered(10.milliseconds, 10.seconds)
-)(implicit timer: Timer = DefaultTimer) extends EnumeratingNamer {
+  mkApi: String => v1.NsApi,
+  labelName: Option[String] = None,
+  backoff: Stream[Duration] = EndpointsNamer.DefaultBackoff
+)(implicit timer: Timer = DefaultTimer)
+  extends Namer {
 
   import EndpointsNamer._
 
-  def lookup(path: Path): Activity[NameTree[Name]]
+  /**
+   * Watch the numbered-port remappings for the service named `serviceName`
+   * in the namespace named `nsName`.
+   * @param nsName the name of the Kubernetes namespace.
+   * @param serviceName the name of the Kubernetes service.
+   * @return an `Activity` containing a `Map[Int, String]` representing the
+   *         port number to port name mappings
+   * @note that the corresponding `Activity` instances are cached so we don't
+   *       create multiple watches on the same Kubernetes objects – meaning,
+   *       if you look up the port remappings in a given (nsName, serviceName)
+   *       pair multiple times, you will always get back the same `Activity`,
+   *       which is created the first time that pair is looked up.
+   */
+  private[this] val numberedPortRemappings =
+    untupled(Memoize[(String, String, Option[String]), Activity[NumberedPortMap]] {
+      // memoize port remapping watch activities so that we don't have to
+      // create multiple watches on the same `Services` API object.
+      case (nsName, serviceName, labelSelector) =>
+        val logEvent = EventLogger(nsName, serviceName)
+        mkApi(nsName)
+          .service(serviceName)
+          .activity(
+            _.map(_.portMappings).getOrElse(Map.empty),
+            labelSelector = labelSelector
+          ) {
+              case (oldMap, v1.ServiceAdded(service)) =>
+                val newMap = service.portMappings
+                logEvent.addition(newMap -- oldMap.keys)
+                oldMap ++ newMap
+              case (oldMap, v1.ServiceModified(service)) =>
+                val newMap = service.portMappings
+                logEvent.addition(newMap -- oldMap.keys)
+                logEvent.deletion(oldMap -- newMap.keys)
+                logEvent.modification(oldMap, newMap)
+                oldMap ++ newMap
+              case (oldMap, v1.ServiceDeleted(service)) =>
+                val newMap = service.portMappings
+                logEvent.deletion(oldMap -- newMap.keys)
+                newMap
+              case (oldMap, v1.ServiceError(error)) =>
+                log.warning(
+                  "k8s ns %s service %s watch error %s",
+                  nsName, serviceName, error
+                )
+                oldMap
+            }
+    })
+
+  private[this] val serviceEndpoints =
```

Comment:
tioli; but this may benefit from an explicit type annotation (very hard to know what the signature of this is by sight)

```scala
 private[this] val serviceEndpoints: String => Option[String =
  ...
```
(or whatever it is)